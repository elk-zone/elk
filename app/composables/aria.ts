export type AriaLive = 'off' | 'polite' | 'assertive'
export type AriaAnnounceType = 'announce' | 'mute' | 'unmute'

const ariaAnnouncer = useEventBus<AriaAnnounceType, string | undefined>(Symbol('aria-announcer'))

export function useAriaAnnouncer() {
  const announce = (message: string) => {
    ariaAnnouncer.emit('announce', message)
  }

  const mute = () => {
    ariaAnnouncer.emit('mute')
  }

  const unmute = () => {
    ariaAnnouncer.emit('unmute')
  }

  return { announce, ariaAnnouncer, mute, unmute }
}

export function useAriaLog() {
  const logs = ref<any[]>([])

  const announceLogs = (messages: any[]) => {
    logs.value = messages
  }

  const appendLogs = (messages: any[]) => {
    logs.value = logs.value.concat(messages)
  }

  const clearLogs = () => {
    logs.value = []
  }

  return {
    announceLogs,
    appendLogs,
    clearLogs,
    logs,
  }
}

export function useAriaStatus() {
  const status = ref<any>('')

  const announceStatus = (message: any) => {
    status.value = message
  }

  const clearStatus = () => {
    status.value = ''
  }

  return {
    announceStatus,
    clearStatus,
    status,
  }
}
