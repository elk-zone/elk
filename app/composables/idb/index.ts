import type { MaybeRefOrGetter, RemovableRef } from '@vueuse/core'
import type { UseIDBOptions } from '@vueuse/integrations/useIDBKeyval'
import { del, get, set, update } from '~/utils/elk-idb'

export interface UseAsyncIDBKeyvalReturn<T> {
  set: (value: T) => Promise<void>
  readIDB: () => Promise<T | undefined>
}

export async function useAsyncIDBKeyval<T>(
  key: IDBValidKey,
  initialValue: MaybeRefOrGetter<T>,
  source: RemovableRef<T>,
  options: Omit<UseIDBOptions, 'shallow'> = {},
): Promise<UseAsyncIDBKeyvalReturn<T>> {
  const {
    flush = 'pre',
    deep = true,
    writeDefaults = true,
    onError = (e: unknown) => {
      console.error(e)
    },
  } = options

  const rawInit: T = toValue<T>(initialValue)

  try {
    const rawValue = await get<T>(key)
    if (rawValue === undefined) {
      if (rawInit !== undefined && rawInit !== null && writeDefaults) {
        await set(key, rawInit)
        source.value = rawInit
      }
    }
    else {
      source.value = rawValue
    }
  }
  catch (e) {
    onError(e)
  }

  async function write(data: T) {
    try {
      if (data == null) {
        await del(key)
      }
      else {
        // IndexedDB does not support saving proxies, convert from proxy before saving
        await update(key, () => toRaw(data))
      }
    }
    catch (e) {
      onError(e)
    }
  }

  const {
    pause: pauseWatch,
    resume: resumeWatch,
  } = watchPausable(source, data => write(data), { flush, deep })

  async function setData(value: T): Promise<void> {
    pauseWatch()
    try {
      await write(value)
      source.value = value
    }
    finally {
      resumeWatch()
    }
  }

  return { set: setData, readIDB: () => get<T>(key) }
}
