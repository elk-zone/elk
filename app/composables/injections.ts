import { InjectionKeyDropdownContext } from '~/constants/symbols'

export function useDropdownContext() {
  return inject(InjectionKeyDropdownContext, undefined)
}
