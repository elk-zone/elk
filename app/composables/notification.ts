import type { mastodon } from 'masto'
import { NOTIFICATION_FILTER_TYPES } from '~/constants'

/**
 * Typeguard to check if an object is a valid notification filter
 * @param obj the object to be checked
 * @returns boolean and assigns type to object if true
 */
export function isNotificationFilter(obj: unknown): obj is mastodon.v1.NotificationType {
  return !!obj && NOTIFICATION_FILTER_TYPES.includes(obj as unknown as mastodon.v1.NotificationType)
}

/**
 * Typeguard to check if an object is a valid notification
 * @param obj the object to be checked
 * @returns boolean and assigns type to object if true
 */
export function isNotification(obj: unknown): obj is mastodon.v1.NotificationType {
  return !!obj && ['mention', ...NOTIFICATION_FILTER_TYPES].includes(obj as unknown as mastodon.v1.NotificationType)
}
