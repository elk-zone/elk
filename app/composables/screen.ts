import { breakpointsTailwind } from '@vueuse/core'

export const breakpoints = useBreakpoints(breakpointsTailwind)

export const isSmallScreen = breakpoints.smallerOrEqual('sm')
export const isMediumOrLargeScreen = breakpoints.between('sm', 'xl')
export const isExtraLargeScreen = breakpoints.smallerOrEqual('xl')
