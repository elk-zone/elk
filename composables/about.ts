import type { BuildInfo } from '~~/types'

export interface Team {
  github: string
  display: string
  twitter: string
  mastodon: string
}

export const elkTeamMembers: Team[] = [
  {
    github: 'antfu',
    display: 'Anthony Fu',
    twitter: 'antfu7',
    mastodon: 'antfu@webtoo.ls',
  },
  {
    github: 'patak-dev',
    display: 'Patak',
    twitter: 'patak_dev',
    mastodon: 'patak@webtoo.ls',
  },
  {
    github: 'danielroe',
    display: 'Daniel Roe',
    twitter: 'danielcroe',
    mastodon: 'daniel@roe.dev',
  },
  {
    github: 'sxzz',
    display: '三咲智子 Kevin Deng',
    twitter: 'sanxiaozhizi',
    mastodon: 'sxzz@webtoo.ls',
  },
].sort(() => Math.random() - 0.5)

export function useBuildInfo() {
  return useAppConfig().buildInfo as BuildInfo
}
