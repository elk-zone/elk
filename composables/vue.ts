import type { ComponentInternalInstance } from 'vue'
import { onActivated, onDeactivated, ref } from 'vue'
import type { ActiveHeadEntry, HeadEntryOptions, UseHeadInput } from '@vueuse/head'
import type { HeadAugmentations } from '@nuxt/schema'
import { useHead } from '#head'

export const isHydrated = ref(false)

export const onHydrated = (cb: () => unknown) => {
  watchOnce(isHydrated, () => cb(), { immediate: isHydrated.value })
}

/**
 * ### Whether the current component is running in the background
 *
 * for handling problems caused by the keepalive function
 */
export function useDeactivated() {
  const deactivated = ref(false)
  onActivated(() => deactivated.value = false)
  onDeactivated(() => deactivated.value = true)

  return deactivated
}

/**
 * ### When the component is restored from the background
 *
 * for handling problems caused by the keepalive function
 */
export function onReactivated(hook: Function, target?: ComponentInternalInstance | null): void {
  const initial = ref(true)
  onActivated(() => {
    if (initial.value)
      return
    hook()
  }, target)
  onDeactivated(() => initial.value = false)
}

// TODO: Workaround for Nuxt bug: https://github.com/elk-zone/elk/pull/199#issuecomment-1329771961
export function useHeadFixed<T extends HeadAugmentations>(input: UseHeadInput<T>, options?: HeadEntryOptions): ActiveHeadEntry<UseHeadInput<T>> | void {
  const deactivated = useDeactivated()
  if (input && typeof input === 'object' && !('value' in input)) {
    const title = 'title' in input ? input.title : undefined
    if (process.server && title) {
      input.meta = input.meta || []
      if (Array.isArray(input.meta)) {
        input.meta.push(
          { property: 'og:title', content: (typeof input.title === 'function' ? input.title() : input.title) as string },
        )
      }
    }
    else if (title) {
      (input as any).title = () => isHydrated.value ? typeof title === 'function' ? title() : title : ''
    }
  }
  return useHead(() => {
    if (deactivated.value)
      return {}
    return resolveUnref(input)
  }, options)
}
