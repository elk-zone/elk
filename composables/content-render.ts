import { TEXT_NODE } from 'ultrahtml'
import type { Node } from 'ultrahtml'
import { Fragment, h, isVNode } from 'vue'
import type { VNode } from 'vue'
import { RouterLink } from 'vue-router'
import { decode } from 'tiny-decode'
import type { ContentParseOptions } from './content-parse'
import { parseMastodonHTML } from './content-parse'
import ContentCode from '~/components/content/ContentCode.vue'
import ContentMentionGroup from '~/components/content/ContentMentionGroup.vue'
import AccountHoverWrapper from '~/components/account/AccountHoverWrapper.vue'

function getTexualAstComponents(astChildren: Node[]): string {
  return astChildren
    .filter(({ type }) => type === TEXT_NODE)
    .map(({ value }) => value)
    .reduce((accumulator, current) => accumulator + current, '')
    .trim()
}

/**
* Raw HTML to VNodes
*/
export function contentToVNode(
  content: string,
  options?: ContentParseOptions,
): VNode {
  let tree = parseMastodonHTML(content, options)

  const textContents = getTexualAstComponents(tree.children)

  // if the username only contains emojis, we should probably show the emojis anyway to avoid a blank name
  if (options?.hideEmojis && textContents.length === 0)
    tree = parseMastodonHTML(content, { ...options, hideEmojis: false })

  return h(Fragment, (tree.children as Node[] || []).map(n => treeToVNode(n)))
}

function nodeToVNode(node: Node): VNode | string | null {
  if (node.type === TEXT_NODE)
    return node.value

  if (node.name === 'mention-group')
    return h(ContentMentionGroup, node.attributes, () => node.children.map(treeToVNode))

  if ('children' in node) {
    if (node.name === 'a' && (node.attributes.href?.startsWith('/') || node.attributes.href?.startsWith('.'))) {
      node.attributes.to = node.attributes.href
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const { href, target, ...attrs } = node.attributes
      return h(
        RouterLink as any,
        attrs,
        () => node.children.map(treeToVNode),
      )
    }
    return h(
      node.name,
      node.attributes,
      node.children.map(treeToVNode),
    )
  }
  return null
}

function treeToVNode(
  input: Node,
): VNode | string | null {
  if (input.type === TEXT_NODE)
    return decode(input.value)

  if ('children' in input) {
    const node = handleNode(input)
    if (node == null)
      return null
    if (isVNode(node))
      return node
    return nodeToVNode(node)
  }
  return null
}

function handleMention(el: Node) {
  // Redirect mentions to the user page
  if (el.name === 'a' && el.attributes.class?.includes('mention')) {
    const href = el.attributes.href
    if (href) {
      const matchUser = href.match(UserLinkRE)
      if (matchUser) {
        const [, server, username] = matchUser
        const handle = `${username}@${server.replace(/(.+\.)(.+\..+)/, '$2')}`
        el.attributes.href = `/${server}/@${username}`
        return h(AccountHoverWrapper, { handle, class: 'inline-block' }, () => nodeToVNode(el))
      }
      const matchTag = href.match(TagLinkRE)
      if (matchTag) {
        const [, , name] = matchTag
        el.attributes.href = `/${currentServer.value}/tags/${name}`
      }
    }
  }
  return undefined
}

function handleCodeBlock(el: Node) {
  if (el.name === 'pre' && el.children[0]?.name === 'code') {
    const codeEl = el.children[0] as Node
    const classes = codeEl.attributes.class as string
    const lang = classes?.split(/\s/g).find(i => i.startsWith('language-'))?.replace('language-', '')
    const code = codeEl.children[0] ? treeToText(codeEl.children[0]) : ''
    return h(ContentCode, { lang, code: encodeURIComponent(code) })
  }
}

function handleNode(el: Node) {
  return handleCodeBlock(el) || handleMention(el) || el
}
