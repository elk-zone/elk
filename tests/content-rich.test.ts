/**
 * @vitest-environment jsdom
 */
/* eslint-disable vue/one-component-per-file */
import type { Emoji } from 'masto'
import { describe, expect, it, vi } from 'vitest'
import { renderToString } from 'vue/server-renderer'
import { format } from 'prettier'
import { contentToVNode } from '~/composables/content-render'

describe('content-rich', () => {
  it('empty', async () => {
    const { formatted } = await render('')
    expect(formatted).toMatchSnapshot()
  })

  it('link + mention', async () => {
    // https://fosstodon.org/@ayo/109383002937620723
    const { formatted } = await render('<p>Happy ðŸ¤— weâ€™re now using <span class="h-card"><a href="https://mas.to/@vitest" class="u-url mention" rel="nofollow noopener noreferrer" target="_blank">@<span>vitest</span></a></span> (migrated from chai+mocha) <a href="https://github.com/ayoayco/astro-reactive-library/pull/203" rel="nofollow noopener noreferrer" target="_blank"><span class="invisible">https://</span><span class="ellipsis">github.com/ayoayco/astro-react</span><span class="invisible">ive-library/pull/203</span></a></p>')
    expect(formatted).toMatchSnapshot()
  })

  it('custom emoji', async () => {
    const { formatted } = await render('Daniel Roe :nuxt:', {
      nuxt: {
        shortcode: 'nuxt',
        url: 'https://media.mas.to/masto-public/cache/custom_emojis/images/000/288/667/original/c96ba3cb0e0e1eac.png',
        staticUrl: 'https://media.mas.to/masto-public/cache/custom_emojis/images/000/288/667/static/c96ba3cb0e0e1eac.png',
        visibleInPicker: true,
      },
    })
    expect(formatted).toMatchSnapshot()
  })

  it('code frame', async () => {
    // https://mas.to/@antfu/109396489827394721
    const { formatted } = await render('<p>Testing code block</p><p>```ts<br />import { useMouse, usePreferredDark } from &#39;@vueuse/core&#39;</p><p>// tracks mouse position<br />const { x, y } = useMouse()</p><p>// is the user prefers dark theme<br />const isDark = usePreferredDark()<br />```</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('code frame 2', async () => {
    const { formatted } = await render('<p><span class=\"h-card\"><a href=\"https://mas.to/@antfu\" class=\"u-url mention\">@<span>antfu</span></a></span> Testing<br />```ts<br />const a = hello<br />```</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('code frame no lang', async () => {
    const { formatted } = await render('<p>```<br />hello world<br />```<br />no lang</p>')
    expect(formatted).toMatchSnapshot()
  })

  it('code frame empty', async () => {
    const { formatted } = await render('<p>```<br /><br />```<br /></p>')
    expect(formatted).toMatchSnapshot()
  })
})

async function render(content: string, emojis?: Record<string, Emoji>) {
  const vnode = contentToVNode(content, { emojis })
  const html = (await renderToString(vnode))
    .replace(/<!--[\[\]]-->/g, '')
  let formatted = ''

  try {
    formatted = format(html, {
      parser: 'html',
    })
  }
  catch (e) {
    formatted = html
  }

  return {
    vnode,
    html,
    formatted,
  }
}

// mocks
vi.mock('vue-router', () => {
  return {
    RouterLink: defineComponent((attrs) => {
      return () => h('a', attrs)
    }),
  }
})

vi.mock('~/composables/dialog.ts', () => {
  return {}
})

vi.mock('~/components/content/ContentCode.vue', () => {
  return {
    default: defineComponent({
      props: {
        code: {
          type: String,
          required: true,
        },
        lang: {
          type: String,
        },
      },
      setup(props) {
        const raw = computed(() => decodeURIComponent(props.code).replace(/&#39;/g, '\''))
        return () => h('pre', { lang: props.lang }, raw.value)
      },
    }),
  }
})

vi.mock('~/components/account/AccountHoverWrapper.vue', () => {
  return {
    default: defineComponent({
      props: ['handle', 'class'],
      setup(_, { slots }) {
        return () => slots?.default?.()
      },
    }),
  }
})
