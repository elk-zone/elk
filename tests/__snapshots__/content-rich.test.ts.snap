// Vitest Snapshot v1

exports[`content-rich > block with backticks 1`] = `"<p><pre class=\\"code-block\\">[(\`number string) (\`tag string)]</pre></p>"`;

exports[`content-rich > block with injected html, with a known language 1`] = `
"<pre>
        <code class=\\"language-js\\">
          &lt;a href=&quot;javascript:alert(1)&quot;&gt;click me&lt;/a&gt;
        </code>
      </pre>
"
`;

exports[`content-rich > block with injected html, with an unknown language 1`] = `
"<pre>
        <code class=\\"language-xyzzy\\">
          &lt;a href=&quot;javascript:alert(1)&quot;&gt;click me&lt;/a&gt;
        </code>
      </pre>
"
`;

exports[`content-rich > block with injected html, without language 1`] = `
"<pre>
        <code>
          &lt;a href=&quot;javascript:alert(1)&quot;&gt;click me&lt;/a&gt;
        </code>
      </pre>
"
`;

exports[`content-rich > code frame 1`] = `
"<p>Testing code block</p><p></p><p><pre class=\\"code-block\\">import { useMouse, usePreferredDark } from &apos;@vueuse/core&apos;
// tracks mouse position
const { x, y } = useMouse()
// is the user prefers dark theme
const isDark = usePreferredDark()</pre></p>"
`;

exports[`content-rich > code frame 2 1`] = `
"<p>
  <span class=\\"h-card\\"
    ><a
      class=\\"u-url mention\\"
      rel=\\"nofollow noopener noreferrer\\"
      to=\\"/webtoo.ls/@antfu\\"
    ></a
  ></span>
  Testing<br />
  <pre class=\\"code-block\\">const a = hello</pre>
</p>
"
`;

exports[`content-rich > code frame empty 1`] = `"<p><pre class=\\"code-block\\"></pre><br></p>"`;

exports[`content-rich > code frame no lang 1`] = `"<p><pre class=\\"code-block\\">hello world</pre><br>no lang</p>"`;

exports[`content-rich > custom emoji 1`] = `
"Daniel Roe
<picture alt=\\":nuxt:\\" class=\\"custom-emoji\\" data-emoji-id=\\"nuxt\\"
  ><source
    srcset=\\"
      https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png
    \\"
    media=\\"(prefers-reduced-motion: reduce)\\" />
  <img
    src=\\"https://media.webtoo.ls/custom_emojis/images/000/000/366/original/73330dfc9dda4078.png\\"
    alt=\\":nuxt:\\"
/></picture>
"
`;

exports[`content-rich > empty 1`] = `""`;

exports[`content-rich > group mention > html 1`] = `
"<p>
  <span class=\\"h-card\\"
    ><a
      class=\\"u-url mention\\"
      rel=\\"nofollow noopener noreferrer\\"
      to=\\"//@pilipinas@lemmy.ml\\"
    ></a
  ></span>
</p>
"
`;

exports[`content-rich > handles formatting from servers 1`] = `
"<h1>Fedi HTML Support Survey</h1>
<p>Does the following formatting come through accurately for you?</p>
<p></p>
<ul>
  <li>This is an indented bulleted list (not just asterisks).</li>
  <li><strong>This line is bold.</strong></li>
  <li><em>This line is italic.</em></li>
</ul>
<ol>
  <li>This list...</li>
  <li>...is numbered and indented</li>
</ol>
<h1>This line is larger.</h1>
"
`;

exports[`content-rich > handles html within code blocks 1`] = `
"<p>
  HTML block code:<br />
  <pre class=\\"code-block\\">
&lt;span class=&quot;icon--noto icon--noto--1st-place-medal&quot;&gt;&lt;/span&gt;
&lt;span class=&quot;icon--noto icon--noto--2nd-place-medal-medal&quot;&gt;&lt;/span&gt;</pre
  >
</p>
"
`;

exports[`content-rich > inline code with link 1`] = `
"<p>
  Inline code with link:
  <code
    >https://api.iconify.design/noto.css?icons=1st-place-medal,2nd-place-medal</code
  >
</p>
"
`;

exports[`content-rich > link + mention 1`] = `
"<p>
  Happy
  <img
    src=\\"/emojis/twemoji/1f917.svg\\"
    alt=\\"ðŸ¤—\\"
    class=\\"iconify-emoji iconify-emoji--twemoji\\"
  />
  weâ€™re now using
  <span class=\\"h-card\\"
    ><a
      class=\\"u-url mention\\"
      rel=\\"nofollow noopener noreferrer\\"
      to=\\"/webtoo.ls/@vitest\\"
    ></a
  ></span>
  (migrated from chai+mocha)
  <a
    href=\\"https://github.com/ayoayco/astro-reactive-library/pull/203\\"
    rel=\\"nofollow noopener noreferrer\\"
    target=\\"_blank\\"
    ><span class=\\"invisible\\">https://</span
    ><span class=\\"ellipsis\\">github.com/ayoayco/astro-react</span
    ><span class=\\"invisible\\">ive-library/pull/203</span></a
  >
</p>
"
`;

exports[`content-rich > plain text 1`] = `
"hello there
"
`;

exports[`editor > transform mentions 1`] = `
"
@elk Hello"
`;
